internal/store/compare.go

```diff
@@ -52,66 +57,181 @@ func Compare(srcDSN, dstDSN string, opts CompareOptions) ([]string, error) {
 		if !ok {
 			return nil, fmt.Errorf("%q table is not found in dst schema", k)
 		}
 
-		// we do a count comparison to save some resources before diving deeper
-		c1, err := srcdb.count(v)
-		if err != nil {
-			return nil, fmt.Errorf("could not count rows of %q: %w", v.TableName, err)
-		}
-		c2, err := dstdb.count(v2)
-		if err != nil {
-			return nil, fmt.Errorf("could not count rows of %q: %w", v2.TableName, err)
-		}
-		if c1 != c2 {
-			mismatchs = append(mismatchs, v.TableName)
-			continue
-		} else if c1 == 0 {
-			continue
+		var c1, c2 int
+		if !opts.SkipCount {
+			// we do a count comparison to save some resources before diving deeper
+			c1, err = srcdb.count(v)
+			if err != nil {
+				return nil, fmt.Errorf("could not count rows of %q: %w", v.TableName, err)
+			}
+			c2, err = dstdb.count(v2)
+			if err != nil {
+				return nil, fmt.Errorf("could not count rows of %q: %w", v2.TableName, err)
+			}
+			if c1 != c2 {
+				fmt.Printf("number of rows did not match for %q(%d, %d)\n", v.TableName, c1, c2)
+				mismatchs[v.TableName] = struct{}{}
+				continue
+			} else if c1 == 0 {
+				continue
+			}
 		}
 
 		remaining := opts.PageSize
 		var cd1, cd2 cursorData
-		var srcCheksum, dstChecksum string
+		var srcChecksum, dstChecksum string
 
+		start := time.Now()
+		var loopCount int
 		// loop until no remaining rows left to calculate checksum
 		for remaining > 0 {
+			loopCount++
 			if cd1.cursors == nil {
 				cd1.limit = remaining
 			}
 			if cd2.cursors == nil {
 				cd2.limit = remaining
 			}
-			srcCheksum, cd1, err = srcdb.checksum(v, cd1)
-			if err != nil {
-				return nil, fmt.Errorf("could not compute src checksum: %w", err)
+
+			// we need to save the previous cursors as they are automatically updated
+			// by the checksum query.
+			prevCd1 := cursorData{
+				cursors: cd1.cursors,
+			}
+			prevCd2 := cursorData{
+				cursors: cd2.cursors,
 			}
 
-			dstChecksum, cd2, err = dstdb.checksum(v2, cd2)
-			if err != nil {
-				return nil, fmt.Errorf("could not compute dst checksum: %w", err)
+			var errSrc, errDst error
+			wg := sync.WaitGroup{}
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				srcChecksum, cd1, errSrc = srcdb.checksum(v, cd1)
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				dstChecksum, cd2, errDst = dstdb.checksum(v2, cd2)
+			}()
+			wg.Wait()
+
+			if errSrc != nil && cd1.limit == 0 {
+				break
+			} else if errSrc != nil {
+				return nil, fmt.Errorf("could not compute src checksums: %w", errSrc)
 			}
 
-			if srcCheksum != dstChecksum {
-				mismatchs = append(mismatchs, v.TableName)
-				continue tableLoop
+			if errDst != nil && cd2.limit == 0 {
+				break
+			} else if errDst != nil {
+				return nil, fmt.Errorf("could not compute dst checksums: %w", errDst)
+			}
+
+			if srcChecksum != dstChecksum {
+				mismatchs[v.TableName] = struct{}{}
+
+				// if verbose flag is set, we print the diff by scanning rows one by one
+				if opts.Verbose {
+				rowLoop:
```

Comment:
Whoop, I had forgotten to publish this comment:

Do we really need the `rowLoop` label? The `break`s and `continue`s I've seen in the loop are not ambiguous.